p2pserver/message/utils/msg_workerpool.go

```diff
@@ -0,0 +1,246 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ /*
+  * https://github.com/valyala/fasthttp
+  * The MIT License (MIT)
+
+  *Copyright (c) 2015-present Aliaksandr Valialkin, VertaMedia
+  *Copyright (c) 2018-present Kirill Danshin
+  *Copyright (c) 2018-present Erik Dubbelboer
+  *Copyright (c) 2018-present FastHTTP Authors
+
+  *Permission is hereby granted, free of charge, to any person obtaining a copy
+  *of this software and associated documentation files (the "Software"), to deal
+  *in the Software without restriction, including without limitation the rights
+  *to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+  *copies of the Software, and to permit persons to whom the Software is
+  *furnished to do so, subject to the following conditions:
+
+  *The above copyright notice and this permission notice shall be included in all
+  *copies or substantial portions of the Software. 
+  */
+
+package utils
+
+import (
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/ontio/ontology-eventbus/actor"
+	"github.com/ontio/ontology/common/log"
+	"github.com/ontio/ontology/p2pserver/message/types"
+	"github.com/ontio/ontology/p2pserver/net/protocol"
+)
+
+type msgJobChanList *[]*msgJobChan
+
+type msgWorkerPool struct {
+	maxWorkerCount    uint
+	maxWorkerIdleTime time.Duration
+	curWorkerCount    uint
+	stopChan          chan struct{}
+	waitingWokers     map[string]msgJobChanList
+	workChanPool      sync.Pool
+	lock              sync.Mutex
+}
+
+type msgJobItem struct {
+	msgPayload *types.MsgPayload
+	p2p        p2p.P2P
+	pid        *actor.PID
+	msgHandler MessageHandler
+}
+
+type msgJobChan struct {
+	lastUseTime time.Time
+	jobChan     chan *msgJobItem
+}
+
+func (this *msgWorkerPool) init() {
+
+	this.waitingWokers = make(map[string]msgJobChanList)
+	this.stopChan = make(chan struct{})
+	this.maxWorkerIdleTime = 10 * time.Second
+}
+
+var msgJobChanCap = func() int {
+	// Use blocking msgJobChan if GOMAXPROCS=1.
+	// This immediately switches receiveMsg to disposeMsg, which results
+	// in higher performance
+	if runtime.GOMAXPROCS(0) == 1 {
+		return 0
+	}
+
+	// Use non-blocking msgJobChan if GOMAXPROCS>1,
+	// since otherwise the receiveMsg caller (Acceptor) may lag accepting
+	// new message if disposeMsg is CPU-bound.
+	return 1
+}()
+
+func (this *msgWorkerPool) start() {
+
+	if this.waitingWokers == nil || this.stopChan == nil {
+		log.Error("[p2p]invalid start invoking, the msg worker pool hasn't been initialized")
+		return
```

Comment:
why return silently? the original implementation is sound, I don't understand why you need to modify it.

Reply:
My original purpose is to avoid the panic once error initialization happen and can't affect other disposings. But the panic has other good aspect. However, I have changed to panic.